xpdan/vend/callbacks/broker.py

```diff
@@ -0,0 +1,320 @@
+import os
+import time as ttime
+from .core import CallbackBase, Retrieve
+from bluesky.utils import ensure_uid
+import numpy as np
+
+
+class BrokerCallbackBase(CallbackBase):
+    """
+    Base class for callbacks which need filled documents
+
+    Parameters
+    ----------
+    fields: Iterable of str
+        Names of data field in an Event
+    db: Broker instance, optional
+        The Broker instance to pull the data from
+    """
+
+    def __init__(self, fields, *, db=None):
+        self.db = db
+        self.fields = fields
+        self.descriptor_dict = {}
+
+    def clear(self):
+        self.descriptor_dict.clear()
+
+    def stop(self, doc):
+        self.clear()
+
+    def descriptor(self, doc):
+        self.descriptor_dict = {doc['uid']: doc}
+
+    def event(self, doc):
+        # the subset of self.fields that are (1) in the doc and (2) unfilled
+        # and (3) external
+        fields = [field for field in self.fields
+                  if (field in doc['data'] and
+                      not doc.get('filled', {}).get(field) and
+                      'external' in self.descriptor_dict[
+                          doc['descriptor']]['data_keys'][field])]
+        if fields:
+            if self.db is None:
+                raise RuntimeError('Either the data must be pre-loaded or '
+                                   'a Broker instance must be provided '
+                                   'via the db parameter of '
+                                   'BrokerCallbackBase.')
+            res, = self.db.fill_events(
+                events=[doc],
+                descriptors=[self.descriptor_dict[doc['descriptor']]],
+                fields=fields)
+            doc['data'].update(**res['data'])  # modify in place
+
+
+class LiveImage(CallbackBase):
+    """
+    Stream 2D images in a cross-section viewer.
```

Comment:
likewise, please take a moment to try and see this comment from the point of view of someone who is not you.  What does this class really do?  Who would use it and why would they use it?  This is then a much more usef docstring moving forward